sphinxcontrib/devhelp/__init__.py

```diff
@@ -109,37 +113,24 @@ def write_toc(node, parent):
         for node in tocdoc.traverse(matcher):  # type: addnodes.compact_paragraph
             write_toc(node, chapters)
 
-        # Index
+        # Index keywords
         functions = etree.SubElement(root, 'functions')
-        index = IndexEntries(self.env).create_index(self)
```

Comment:
Why removing general index? domain indices are different from general index. They are not alternative.

Reply:
Because this index refers to the indexing of [keywords](https://github.com/sphinx-doc/sphinxcontrib-devhelp/blob/f8ca4ed86af2146712c5006a6fb0920a0eb34dbd/sphinxcontrib/devhelp/__init__.py#L117-L124), references to path description of modules, functions or attributes, this helps in the search for such information in the search of GNOME DevHelp. The general index has already been made [here](https://github.com/sphinx-doc/sphinxcontrib-devhelp/blob/f8ca4ed86af2146712c5006a6fb0920a0eb34dbd/sphinxcontrib/devhelp/__init__.py#L112-L114)

I thought about keeping IndexEntries, but it did not give me access to other information like, the [type](https://github.com/sphinx-doc/sphinxcontrib-devhelp/blob/f8ca4ed86af2146712c5006a6fb0920a0eb34dbd/sphinxcontrib/devhelp/__init__.py#L125) of the keyword (eg: class, module, attribute). Unless I use regular expressions like in [qthelp](https://github.com/sphinx-doc/sphinxcontrib-qthelp/blob/1307e1347a722ec0c23e000acf7d310fe920e389/sphinxcontrib/qthelp/__init__.py#L39-L40).

Reply:
>The general index has already been made here

No, this is toctree (table of contents), not a general index. The general index contains list of index entries marked up by `index` directive and so on. I think this removal means devhelp does not support `index` directive. Is my understanding correct?

Reply:
@tk0miya A Yes, disregard this statement. I messed with the terms.

But this change still maintains an index (it's done by Devhelp [here](https://github.com/sphinx-doc/sphinxcontrib-devhelp/blob/f8ca4ed86af2146712c5006a6fb0920a0eb34dbd/sphinxcontrib/devhelp/__init__.py#L117)), but now it's powered by the domains of documentation and domains that address the requirements of "functions" in the devhelp XML (which are, name, [type](https://github.com/sphinx-doc/sphinxcontrib-devhelp/blob/f8ca4ed86af2146712c5006a6fb0920a0eb34dbd/sphinxcontrib/devhelp/__init__.py#L121-L125), and link).

Reply:
@tk0miya is there any way to know the "type" (eg class, module, function) with index? I used domains because he gave me this information. As for the Index, I would have to use regexp taking each value in the title

Reply:
Are you saying about the objtype of python domain, right? If so, you can see the list of objtypes for python domain here:
https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/domains/python.py#L1154-L1164

Note: The objtype is extensible. So 3rd party extensions can add their own objtypes additionally.

Reply:
Yes, that's about it, but not just in Python, but any available domain that is built for devhelp.

The problem here is that `IndexEntries`¹ already delivers the ready index with these values inserted in the titles, for example, _(Point **attribute**)_, "**_attribute_**" being the value I need. To solve this I would have to use regular expression to get these objtype values, just like qthelp² does. Would that make more sense?

I am basing myself on the new specification³, and for this item (keyword⁴) it asks for the type of keyword. 

[1] https://github.com/sphinx-doc/sphinx/blob/7ecf6b88aa5ddaed552527d2ef60f1bd35e98ddc/sphinx/environment/adapters/indexentries.py#L30
[2] https://github.com/sphinx-doc/sphinxcontrib-qthelp/blob/master/sphinxcontrib/qthelp/__init__.py#L201-L202
[3] https://gitlab.gnome.org/GNOME/gtk-doc/-/blob/master/devhelp2.xsd#L139
[4] https://gitlab.gnome.org/GNOME/gtk-doc/-/blob/master/devhelp2.xsd#L94

Reply:
The label is generated by `get_index_text()`.
https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/domains/python.py#L929

But it is easily changed by translation. So it's not a good way to do that. I think qthelp will not work if document is translated (it's bug!).

Using `PythonDomain.objects` is much better to list documented python objects up. And it contains objtype info
https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/domains/python.py#L1200
https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/domains/python.py#L71

You can access it via `env.get_domain('python')`:

```
pydomain = env.get_domain('python')
for name, entry in pydomain.objects:
    print("%s: %s\n" % (name, entry.objtype))
```

Reply:
I understand this approach ... but is it very different from the previous¹ patch?

```py
for domain_name in sorted(self.env.domains):
    for domain_data in self.env.domains[domain_name].get_objects():
        name, _, _type, docname, anchor, _ = domain_data
        link = '{docname}.html#{anchor}'.format(docname=docname, anchor=anchor)
        etree.SubElement(functions, 'keyword', name=name, link=link, type=_type)
```

Because that way, I will still need to take the "domains" (self.env.domains) available and make a loop on it and go calling self.env.get_domain

```py
for domain in sorted(self.env.domains):
    pydomain = self.env.get_domain(domain)
    for obj in pydomain.get_objects():
        name, _, _type, docname, anchor, _ = obj
```

Unless I'm being very inattentive, but both are "alike" and solve the problem.

[1] - https://github.com/gutierri/sphinxcontrib-devhelp/commit/a14220561fc76401c374762f0c0869ba9e3929bc?branch=a14220561fc76401c374762f0c0869ba9e3929bc&diff=unified#diff-8b855eba53d212c601c6a27b04e90e916faf2e276ac9d49b7fc66e80dbfaab86L115-L122

Reply:
Sorry, I still don't understand your goal. I thought you'd like to get the type of python object. So I suggested you to use the python domain to get them all. But it seems you access all domains. For example, sphinx also has math and versionchanged domains. Would you really like to get the types of these objects too?

Reply:
No problem... my English is in "wip", then the communication is sometimes complicated :P

Yes, thinking that the idea of this part of the devhelp is to bring the domains and their types. I believe that any problem can be adjusted in the future, but I have used this approach and works. 

Can I return the previous patch then and merge work?

Reply:
I returned to the previous patch. I believe that now it is in accordance with the objective (with the exception of the error that occurred when running the tests)

Reply:
Do you know the difference between the objects in the domains and the general index? They're similar, but different things. So you need to know to implement this feature. It seems the new index does not contain the entries in the general index.

The former one, the objects in the domains are the list of "objects" for each domain. The types of object are determined by domains. For example, The python domain has function, class, attribute, exception, etc. as object types. The reST domain has directive, directive option, and role as object types. These objects can be defined via its directives like `py:function` and `rst:directive:`.

The latter one, the general index is a collection of general index-entries. The index entries can be defined via `index` directive. And they're also defined via domain's directives. So you can see index entries both defined in `index` directive and domain's directive in the general index. The `IndexEntries` class is a manipulator class of the general index.

It seems you're going to replace the general index by the objects in domains. It means the index entries defined by `index` directive will not be shown in the devhelp2 index. Is it expected? I need to know the ideal form of devhelp2 before discussing the implementation.

Reply:
Yes, the general index makes up the pages and the domain maps the documented "tokens" and makes reference to the documentation.

The general index has already been generated on line 94-115, the "functions" referring to "tokens" and "keywords" and each one has a type, so it is necessary to use domains and not a general index generic.

I am putting some examples with screenshots of this patch in use here and its behavior that is with the correct tables of contents and working and also the refined search with the use of the domains.

![2021-05-04-13--06--28](https://user-images.githubusercontent.com/1111356/117035654-f5b41580-acda-11eb-9f16-aaac6b9976ec.png)

![2021-05-04-13--07--11](https://user-images.githubusercontent.com/1111356/117036027-60fde780-acdb-11eb-955d-468cb41fe07b.png)

Reply:
It seems you're confused. The "general index" you called is not a real "general index". In Sphinx, it's called as ToC (table-of-contents). In Sphinx, the "general index" is a list of keywords. See https://www.sphinx-doc.org/en/master/genindex.html. It was originally generated here (via the `IndexEntries` class).

It contains the keywords in domains. Addition to them, it also contains the keywords registered manually via `index` directive and role. Some extensions also can add it.

Please check your code output the keyword defined by `index` directive.

Reply:
yes, I was really confused when explaining, sorry about that. But the point is that I need to know the domains contained in the documentation to make the construction an index of keywords (which would be the modules, variable functions and etc... what is the self.domain_indicie). IndexEntry does this, but not as I need it, it returns a ready-made string. similar to this: "MyMasterClass (class in mymodule)"

For devhelp2 I need to add an XML attribute, which is the type to fill this value, I decided to use self.domain_indices.

It was clear? And sorry for the long no update.